text/libobject-refactor.md

```diff
@@ -0,0 +1,197 @@
+  register it with `Lib.add_leaf`. `Lib`'s job is mainly to sit below
+  all clients that can register objects, and track the (ephemeral)
+  list of objects that have been declared while a module / library is
+  built.
+
+- when a module or section is closed, `Declaremods` will call `Lib` to
+  collect the corresponding objects, and in some cases they will be
+  re-declared if they survive the module / section.
+
+- at the end of a library compilation, `Library` will call
+  `Declaremods.end_library` to grab the list of objects to persist in
+  the `.vo` file.
+
+## Loading
+
+- when a new library or module is `Required`, `Declaremods` will
+  sequentially "load" the objects, basically performing the
+  side-effects declared in their interface
+
+- the objects are then stored in a internal map, as their interface
+  may need to be called when importing for example
+
+- users of `Declaremod` have to be careful about the state
+
+# Logical and meta-logical objects
+
+A common discussion topic is about the role that objects registered
+with the kernel play. Indeed, some objects present in a `.vo` can be
+understood as "logical" objects, as they will be registered with the
+kernel.
+
+A large set of objects are meta-logical: they are not registered with
+the kernel, but still essential to Coq operation. Such objects are
+notations, instance declarations, hints, etc...
+
+Indeed it is a common source of discussion where the data for logical
+objects should be stored. Some data will be stored invariant in the
+kernel, but some other data, like documentation or location, has to
+live outside it.
+
+# Part I: common meta-data
+
+For this part, we propose to decorate the `Libobject.t` type with a
+`Data.t` structure. This approach is modelled after the approach used
+in `Declare` for constant information, which has proven successful in
+reducing code duplication, uniformizing the API, and making it more
+extensible (see for example per-definition typing flags).
+
+In this preliminary proposal, `Data.t`, will contain:
+
+- `loc`: Location of the object, if it is coming from a source file
+- `doc`: String containing the documentation of the object, to be extended
+- `name`: Name of the object, as an `Id.t`, if not anonymous
+- `category`: Already exiting in `Libobject`, used for import, could be also used by `Search`.
+
+A draft implementation is available in PR (number to be added ASAP)
+
+This changes affect the lifetime in the following way:
+
+## Creation
+
+Clients must provide the metadata when registering the objects.
+
+## Loading
+
+`Declaremods` already stores the objects in a map indexed by their
+prefix, we propose to:
+
+- create a reverse map or maps to allow clients to query the `Data.t`
+  efficiently. In particular we could have reverse maps for
+  categories, names, locations. For now we implement a reverse map for
+  names, but more options could be desirable after more experimentation.
```

Comment:
Does this refer to a branch somewhere or is more like "for now we propose to start with a reverse map for names"?

Reply:
This is implemented in the prototype branch, cleaning it up to submit as we speak.

Reply:
PR done, you can use (hack hack) `Print Graph` to print the reverse map